_posts/2023-04-10-opensource-perf.markdown

```diff
@@ -0,0 +1,129 @@
+
+When operating an email server, performance is a concern, as email is a massively used application and is generally considered to be critical. Often, with even a medium-sized deployment, handling over 1,000 requests per second is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explain](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.
+
+Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from Elasticsearch 7.10 to OpenSearch for licensing reasons (as an Apache project, we must use licenses compatible with the Apache License 2.0). We realized the OpenSearch requests were slower and started investigating why.
+
+## Identifying the performance issue
+
+Flame graphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flame graphs, we were able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flame graph and found that frequent SPI lookups were causing the performance issue.
+
+Here is the flame graph we used to diagnose this issue. It was taken with [async-profiler](https://github.com/jvm-profiling-tools/async-profiler) from within an Apache James container:
+
+`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`
+
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }
+
+Then we searched for `"opensearch"` using the search widget in the upper-left corner of the flame graph:
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch client threads"/>{: .img-fluid }
+
+-   The client thread (on the right) with the httpclient event loop.
```

Comment:
It's not immediately clear what these two bullet points are connected to. They would normally be introduced by a brief sentence ending in a colon. Should they just be a short paragraph instead of bullets?

Reply:
Makes sense! I moved it to the above the graph since it represents what is shown in the graph.